Updated READ.md & Added keyboard shortcuts

mebinthattil opened the pull request:
## Summary
Add two new features to the repository:
1. Update `README.md`
2. Add keyboard shortcuts  

---

## What it Solves

### README  
- Update the download link to point to the direct download link instead of searching it up.  

### Keyboard Shortcuts  
- Add keyboard shortcuts for common actions, including:  
  1. `Ctrl+S` -> Speaks the current text
  2. `Ctrl+1` -> Switch to speak mode
  3. `Ctrl+2` -> Switch to chatbot mode
  4. `Ctrl+3` -> Switch to voice mode
  5. `Ctrl+M` -> Cycle through mouth types

- Added UI indicators so users can discover these shortcuts:
![Shortcut hint in speak](https://github.com/user-attachments/assets/8d20bebb-118b-44bd-9b15-fb08fa48f4d4)


---

## Testing  
Tested the changes locally to ensure that all shortcuts work as intended, regardless of screen focus. Confirmed that shortcuts function correctly under all conditions.

The following requirements have been tested:
- [x] Activity must start  
- [x] Activity must refresh entire display when focus is restored  
- [x] Activity must behave predictably (except where randomness is designed)  
- [x] Activity must save data to journal  
- [x] Activity must restore the saved data from journal (click on the journal entry)  
- [x] Other activities must be able to use saved data, if it is declared with the relevant content type  
- [x] Every coded feature should either work properly, or be removed if it cannot be fixed  
- [x] Collaboration support, if present, must function properly between two or more systems  
- [x] Activity should not consume all available battery power (e.g. pygame clock rate too high)  
- [x] Activity should not contain any security vulnerabilities  
- [x] Activity should not reveal personal information  

Please let me know if anything needs to be changed / updated.

quozl commented:
Please never close a pull request unnecessarily.  You have split away from the other discussion, and it looks like you are trying to hide it.  Of course you aren't, but there's no reason to have closed.  You could have reset your branch to ec299b7 and pushed it and your other pull request would be updated.

The older pull request https://github.com/sugarlabs/speak/pull/35

Our guidance on branching you have not followed;
* https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#branching

Our guidance on commit messages you have not followed;
* https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#making-commits

GitHub correctly abbreviates your ec299b7 commit message when rendering the pull request pages.  The ellipsis shows you have not kept to common commit message formatting, which our guidance reinforces.

Please fix the commit messages.

mebinthattil commented:
Going forward I shall make sure to strictly abide by the contribution docs. The contribution docs mention that I need to create a separate branch and create a PR from that to the master branch of the speak repo. 

But since this current PR is on the **master** branch of my fork does that mean that I have to close this PR and create another PR from a separate branch? Is there a way I can change the source branch for a PR?

I apologize for any inconvenience caused and assure you that I will take steps to prevent such mistakes from occurring in the future.

quozl commented:
I've never had to do it, because I always get it right.  :grinning: 

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request looks interesting.

mebinthattil commented:
That article mentions how the base repository can be changed, but in my case I want to change the source repo (which I think cannot be edited).

Changing base repo would mean that I could merge this PR into another fork from the sugar speak repo. 

But what I need to do is to merge this PR to the **master of speak repo**, but from a **branch** on **my fork**.

So does this leave me with no other option other than closing this PR? Again sorry for this, will make sure not to make these mistakes again.

quozl commented:
Hmm, thanks, let me know what you find.

quozl commented:
You should learn how to rewrite your commits and then force push, that way only this pull request is updated.

mebinthattil commented:
@quozl I have made changes to my commit message to align with the contribution standards. I hope I have done it right this time. 
I wont be able to change the branch I’m creating this PR from as that is immutable. I hope that is fine.

quozl commented on README.md:
```diff
@@ -10,17 +10,19 @@ How to use?
 Speak is part of the Sugar desktop and is often included.  Please refer to;
 
-* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
+* [Try Online](https://try.sugarizer.org/),
```
Sugarizer is not Sugar desktop, so inappropriate to promote here.  Will yield experience confusion.

mebinthattil replied:
Noted, I have deleted it.

quozl commented on README.md:
```diff
@@ -10,17 +10,19 @@ How to use?
 
-* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
+* [Try Online](https://try.sugarizer.org/),
+* [Experience Sugar](https://www.sugarlabs.org/#try),
```
Experience Sugar link doesn't help in any way related to Speak activity.

mebinthattil replied:
Noted, removed it.

quozl commented on README.md:
```diff
@@ -10,17 +10,19 @@ How to use?
+* [Try Online](https://try.sugarizer.org/),
+* [Experience Sugar](https://www.sugarlabs.org/#try),
 * [How to use Sugar](https://help.sugarlabs.org/),
-* [Download Speak using Browse](https://v4.activities.sugarlabs.org/), search for `Speak`, then download, and;
```
You have removed my " and;" which signals second last list item, under my Australian Style Guide.

quozl commented on README.md:
```diff
@@ -10,17 +10,19 @@ How to use?
-* [Download Speak using Browse](https://v4.activities.sugarlabs.org/), search for `Speak`, then download, and;
+* [Download Speak using Browse](https://v4.activities.sugarlabs.org/app/vu.lux.olpc.Speak.html), then download,
 * [How to use Speak](https://help.sugarlabs.org/en/speak.html).
+* [Sugar Labs Wiki](https://wiki.sugarlabs.org/go/Welcome_to_the_Sugar_Labs_wiki)
```
I don't think link helps with Speak.  Also, why doesn't wiki.sugarlabs.org by itself not load?

mebinthattil replied:
Noted, removed it. Also the sugar labs wiki website seems to be down right now.

quozl commented on README.md:
```diff
@@ -10,17 +10,19 @@ How to use?
 How to upgrade?
 ===============
-
+    
```
Change not required.

mebinthattil replied:
Removed it, it was not intentional.

quozl commented on README.md:
```diff
@@ -10,17 +10,19 @@ How to use?
 On Sugar desktop systems;
 * use [My Settings](https://help.sugarlabs.org/en/my_settings.html), [Software Update](https://help.sugarlabs.org/en/my_settings.html#software-update), or;
-* use Browse to open [v4.activities.sugarlabs.org](https://v4.activities.sugarlabs.org/), search for `Speak`, then download.
+* use Browse to open [https://v4.activities.sugarlabs.org/app/vu.lux.olpc.Speak.html](https://v4.activities.sugarlabs.org/app/vu.lux.olpc.Speak.html),then download.
```
Missing space; deletes were too much.

mebinthattil replied:
Made changes.

quozl commented on README.md:
```diff
@@ -53,3 +55,4 @@ The `not-gstreamer1` branch is a backport of features and bug fixes
 from the `master` branch for ongoing maintenance of the activity on
 Fedora 18 systems which don't have well-functioning GStreamer 1
 packages.
+
```
Change not required.

mebinthattil replied:
Deleted, change was not intentional.

quozl commented on activity.py:
```diff
@@ -937,6 +937,41 @@ def _entry_key_press_cb(self, combo, event):
             return True
+        
+        # Adding keyboard shortcuts for common actions
+        elif event.state & Gdk.ModifierType.CONTROL_MASK:
```
Shortcuts won't work if the keyboard focus is elsewhere; add them to the activity window as shortcuts for the buttons if they aren't already in place.

mebinthattil replied:
I have added them to the activity windows, thanks for pointing that out.

quozl reviewed (changes requested):
Your commit messages did not explain why changes were made.

mebinthattil commented:
@quozl I have made the requested changed, kindly review it. Thanks

quozl commented on README.md:
```diff
@@ -12,15 +12,19 @@ Speak is part of the Sugar desktop and is often included.  Please refer to;
 
+The speak activity can also be tried online. Please refer to;
+
+* [Try Speak Online](https://try.sugarizer.org/), log in, or create an account; then use the speak activity.
```
Sugarizer's speak has diverged, which is expected.  I don't think it is appropriate to reference it, because it suggests that the experience is the same, which it isn't.  Long term we could use the Sugarizer Speak inside Sugar in place of the older Python version.

mebinthattil replied:
I understand that there are differences in the sugar speak and the sugarizer speak, but the fundamentals of what the speak activity is, is still the same right?

Providing a link to try speak online I think offers a path of very low friction for someone to quickly try out speak before actually installing it on their machine. It also gives a chance to try out the activity if the person does not have the environment to run the activity (windows, mac).

If you are still of the opinion that we must not keep this link, I shall remove it. Thanks

quozl replied:
Speak in Sugarizer has diverged features from Speak in Sugar, and that is a normal part of software evolution, led by @llaske.  I don't want developers of Speak in Sugar to feel any desire to track those changes; instead of tracking the changes we would replace the Python source code with the JavaScript source code converting this activity from Python to JavaScript implementation.  This makes more sense for future maintenance, as the number of maintainers falls.

However, we currently have a GSoC idea [Refactor the chatbot in the Speak Activity to use gen-AI](https://github.com/sugarlabs/GSoC/blob/master/Ideas-2025.md#Refactor-the-chatbot-in-the-Speak-Activity-to-use-gen-AI) which would be harder to implement if the activity was replaced with JavaScript, so the replacement is certainly not planned at this time.

@chimosky your opinion?

chimosky replied:
I agree with Quozl, referencing the Speak from Sugarizer isn't great as it's diverged from the one in Sugar.

Referencing it here creates expectations that'll lead to problems.

quozl commented on activity.py:
```diff
@@ -1400,3 +1451,4 @@ def set_widget(self, widget):
         widget.show()
 
     widget = GObject.Property(getter=get_widget, setter=set_widget)
+
```
Unintended change?

mebinthattil replied:
Yes, this is unintentional. I keep removing the trailing empty line at the end, but every time I commit, it seems to add a new line at the end. I will try and change that. Thanks

P.S: I’m using VS Code as my IDE

quozl replied:
You might try interactive git add.  Quick search yields https://stackoverflow.com/questions/34730585/vs-code-how-to-stage-and-commit-individual-changes-in-a-single-file

quozl reviewed (changes requested):
Reviewed again, thanks.

mebinthattil commented:
@quozl Sorry for getting back to this a little late.
I have made the requested changes, which were:
- Removing the sugarizer link from the `README.md`, as it would cause confusion,
- Removed the trailing new line at the end of the `activity.py` file.

If there is any more feedback or anything that needs to be changed, kindly let me know. Thanks!

chimosky commented on README.md:
```diff
@@ -12,15 +12,15 @@ Speak is part of the Sugar desktop and is often included.  Please refer to;
 * [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
 * [How to use Sugar](https://help.sugarlabs.org/),
-* [Download Speak using Browse](https://v4.activities.sugarlabs.org/), search for `Speak`, then download, and;
+* [Download Speak using Browse](https://v4.activities.sugarlabs.org/app/vu.lux.olpc.Speak.html), then download, and;
```
From your commit message, the download link doesn't have any problems because the instructions clearly state how to get the activity using the download link, I'm fine with changing the link to the direct page but then don't say it was pointing to the old website because it's not.

What old website are you referring to?

mebinthattil replied:
You're right, its pointing to the same page itself, I can't seem to recollect why I wrote that. Sorry, I shall change that. Thanks

mebinthattil replied:
I figured out why I had mentioned that I changed the **old** website link. 
When I started working on this issue, #32 had not been merged, so the link was still pointing to `https://activities.sugarlabs.org/` instead of `https://v4.activities.sugarlabs.org/` (new website). 
Later once that PR was merged, I had git pulled, so thats why I thought I changed to the new website, but turns out it had already been done.
Hope that clears it. :)

chimosky replied:
Alright then.

mebinthattil replied:
Yes, I had removed that from my commit message. Is it not visible?

chimosky commented on README.md:
```diff
@@ -12,15 +12,15 @@ Speak is part of the Sugar desktop and is often included.  Please refer to;
 On Sugar desktop systems;
 * use [My Settings](https://help.sugarlabs.org/en/my_settings.html), [Software Update](https://help.sugarlabs.org/en/my_settings.html#software-update), or;
-* use Browse to open [v4.activities.sugarlabs.org](https://v4.activities.sugarlabs.org/), search for `Speak`, then download.
+* use Browse to open [https://v4.activities.sugarlabs.org/app/vu.lux.olpc.Speak.html](https://v4.activities.sugarlabs.org/app/vu.lux.olpc.Speak.html), then download.
```
Same issue as above.

mebinthattil replied:
Yes, shall fix it, thanks!

chimosky commented on activity.py:
```diff
@@ -1399,4 +1450,4 @@ def set_widget(self, widget):
         widget.show()
 
-    widget = GObject.Property(getter=get_widget, setter=set_widget)
+    widget = GObject.Property(getter=get_widget, setter=set_widget)
```
Please remove the no new line at EOF.

chimosky reviewed (commented):
You've added new shortcuts which I don't see why, and you've also not tried to improve the documentation to actually show that these shortcuts exist, how exactly would anyone figure these out?

mebinthattil commented:
> You've added new shortcuts which I don't see why, and you've also not tried to improve the documentation to actually show that these shortcuts exist, how exactly would anyone figure these out?

Right, where should I update the documentation for this. Would adding this to the `README.md` make sense? Or should I create some kind of a hint / indicator message within the activity itself? Thanks

mebinthattil commented:
@chimosky I have made the requested change which were:
- Updating commit message, as it incorrectly mentioned that the link points to the new website, which was a mistake on my end. It just provides the direct download link
- Resolved the issue with the EOF new line.

Let me know if anything else needs to be changed / fixed. Thanks!

quozl commented:
* [ ] in fc83a1116a74cee43a567307654e573bd40d95b8 the link text and surrounding text for the second link is complex, please rewrite,
* [ ] in 357e5a0882c66ec7eb01d23788d861bbafb68ec8 the shortcuts are added globally instead of to the widgets where they would usually be added as accelerators and tooltips, consequently there are no tooltips, and as @chimosky correctly pointed out, no way to discover them short of reading source code or documentation.

chimosky commented:
> > You've added new shortcuts which I don't see why, and you've also not tried to improve the documentation to actually show that these shortcuts exist, how exactly would anyone figure these out?
> 
> Right, where should I update the documentation for this. Would adding this to the `README.md` make sense? Or should I create some kind of a hint / indicator message within the activity itself? Thanks

The Readme is one, a better place would be the activity help button, if there's one.

mebinthattil commented:
> * [x]  in [fc83a11](https://github.com/sugarlabs/speak/commit/fc83a1116a74cee43a567307654e573bd40d95b8) the link text and surrounding text for the second link is complex, please rewrite,
> * [x]  in [357e5a0](https://github.com/sugarlabs/speak/commit/357e5a0882c66ec7eb01d23788d861bbafb68ec8) the shortcuts are added globally instead of to the widgets where they would usually be added as accelerators and tooltips, consequently there are no tooltips, and as @chimosky correctly pointed out, no way to discover them short of reading source code or documentation.

@quozl I have now implemented the same shortcuts using GTK accelerators, and removed the globally declared function. I hope this is the right way to do it. 
`README.md` has been updated as per feedback.

@chimosky I created a keyboard shortcut hint, which shows when hovering over the icons as shown below:
![Shortcut hint in speak](https://github.com/user-attachments/assets/3007ffa6-7bbb-440b-8a31-f17eefff1676)

Do let me know your thoughts on this, and if there is anything that needs to be changed / improved upon. Thanks!

quozl commented:
Thanks.  Looked at your change up to 3dca125.

* [ ] keyboard shortcuts do not have the words "keyboard shortcut", hover the mouse over the stop button to see how it should look,
* [ ] your code changes are far too complicated; the Activity class already has an accel_group, and you only need to assign to widget.props.accelerator, which is done automatically by ToolButton, which is a subclass of ToolbarButton ... if the right arguments are provided.

Suggest cloning several other activities and reading the variety of code to see how best to do this.

chimosky commented on activity.py:
```diff
@@ -245,7 +249,7 @@ def __init__(self, handle):
         self._chat.show_all()
         self._notebook.append_page(self._chat, Gtk.Label(''))
-
+        
```
Also look at the diff of your change before committing to avoid changes like this.

quozl commented:
Your two new merge commits are not necessary and detract from the review.  You can remove them by using `git reset` to return to [a6d0b5d](https://github.com/sugarlabs/speak/pull/36/commits/a6d0b5d5de1ea9b4f94bf36bc6a9979348a3d51f) or use rebase.

mebinthattil commented:
I apologise, this was not intentional. I was working on another feature and I synced my fork. 

Because this PR was created on my master branch, it automatically added that.

Do you mind if I close this PR for now and open it back when I intend on picking this back up? I'm planning on adding things to my fork of speak, so it doesn't make sense that a PR from my master is still open.

I realise that opening a PR from my master branch was a bad idea.

chimosky commented:
I'll close this now, you can open a new one if you ever need to.
